OpenRA.Mods.Common/Widgets/CheckboxWidget.cs

```diff
@@ -53,38 +53,37 @@ public override void Draw()
 		{
 			var disabled = IsDisabled();
 			var font = Game.Renderer.Fonts[Font];
+			var hover = Ui.MouseOverWidget == this;
 			var color = GetColor();
 			var colordisabled = GetColorDisabled();
-			var bgDark = GetContrastColorDark();
-			var bgLight = GetContrastColorLight();
-			var rect = RenderBounds;
 			var text = GetText();
-			var textSize = font.Measure(text);
-			var check = new Rectangle(rect.Location, new Size(Bounds.Height, Bounds.Height));
-			var baseName = IsHighlighted() ? "checkbox-highlighted" : "checkbox";
-			var state = WidgetUtils.GetStatefulImageName(baseName, disabled, Depressed && HasPressedState, Ui.MouseOverWidget == this);
-
-			WidgetUtils.DrawPanel(state, check);
+			var rect = new Rectangle(RenderBounds.Location, new Size(Bounds.Height, Bounds.Height));
 
-			var topOffset = font.TopOffset;
-			var position = new float2(rect.Left + rect.Height * 1.5f, RenderOrigin.Y + (Bounds.Height - textSize.Y - topOffset) / 2);
+			DrawBackground(Background, rect, disabled, Depressed, hover, IsHighlighted());
 
+			var textPosition = new float2(RenderBounds.Left + RenderBounds.Height * 1.5f, RenderOrigin.Y + (Bounds.Height - font.Measure(text).Y - font.TopOffset) / 2);
 			if (Contrast)
-				font.DrawTextWithContrast(text, position,
-					disabled ? colordisabled : color, bgDark, bgLight, 2);
+				font.DrawTextWithContrast(text, textPosition,
+					disabled ? colordisabled : color, GetContrastColorDark(), GetContrastColorLight(), 2);
 			else
-				font.DrawText(text, position,
+				font.DrawText(text, textPosition,
 					disabled ? colordisabled : color);
 
-			if (IsChecked() || (Depressed && HasPressedState && !disabled))
+			if (getCheckmarkImageCache == null)
 			{
-				var checkType = GetCheckType();
-				if (HasPressedState && (Depressed || disabled))
-					checkType += "-disabled";
-
-				var offset = new float2(rect.Left + CheckOffset, rect.Top + CheckOffset);
-				WidgetUtils.DrawSprite(ChromeProvider.GetImage("checkbox-bits", checkType), offset);
+				getCheckmarkImageCache = new CachedTransform<(string, bool, bool), CachedTransform<(bool, bool, bool, bool), Sprite>>(
+					((string CheckType, bool Highlighted, bool Checked) args) =>
+					{
+						var variantImageCollection = args.Highlighted ? "checkmark-highlighted-" : "checkmark-" + args.CheckType;
+						var variantBaseName = args.Checked ? "checked" : "unchecked";
+						return WidgetUtils.GetCachedStatefulImage(variantImageCollection, variantBaseName);
+					});
```

Comment:
Cant we do this once in the constructor?

Reply:
Yes we can, good catch